src/main/scala/devices/spi/SPIPhysical.scala

```diff
@@ -59,11 +98,39 @@ class SPIPhysical(c: SPIParamsBase) extends Module {
     Mux(fmt.endian === SPIEndian.MSB, data, Cat(data.toBools))
 
   val rxd = Cat(io.port.dq.reverse.map(_.i))
-  val samples = Seq(rxd(1), rxd(1, 0), rxd)
+  val rxd_delayed = Vec(Seq.fill(io.port.dq.size)(false.B))
+  //0.U(io.port.dq.size.W).toBools
+  //Adding fine-granularity delay buffers on the received data
+
+  val fine_grain_delay_0 = Module(new BlackBoxDelayBuffer())
+  val fine_grain_delay_1 = Module(new BlackBoxDelayBuffer())
+  val fine_grain_delay_2 = Module(new BlackBoxDelayBuffer())
+  val fine_grain_delay_3 = Module(new BlackBoxDelayBuffer())
+  
+  fine_grain_delay_0.io.in := rxd(0)
+  fine_grain_delay_0.io.sel := io.ctrl.extradel.fine
+  rxd_delayed(0) := fine_grain_delay_0.io.mux_out
+  
+  fine_grain_delay_1.io.in := rxd(1)
+  fine_grain_delay_1.io.sel := io.ctrl.extradel.fine
+  rxd_delayed(1) := fine_grain_delay_1.io.mux_out
+  
+  fine_grain_delay_2.io.in := rxd(2)
+  fine_grain_delay_2.io.sel := io.ctrl.extradel.fine
+  rxd_delayed(2) := fine_grain_delay_2.io.mux_out
+  
+  fine_grain_delay_3.io.in := rxd(3)
+  fine_grain_delay_3.io.sel := io.ctrl.extradel.fine
+  rxd_delayed(3) := fine_grain_delay_3.io.mux_out
+  ////////
+  val rxd_fin = rxd_delayed.asUInt
+  val samples = Seq(rxd_fin(1), rxd_fin(1, 0), rxd_fin)
 
   val buffer = Reg(op.data)
   val buffer_in = convert(io.op.bits.data, io.ctrl.fmt)
-  val shift = if (c.sampleDelay > 0) setup_d || (sample_d && stop) else sample_d
+  //val shift = if (io.ctrl.extradel.coarse > 0.U) setup_d || (sample_d && stop) else sample_d
```

Comment:
do we need to leave the commented code here?

Reply:
WIll remove.